### PR TITLE
LPD-28413 (UserERC) and LPD-28412 (GroupERC and ScopeGroupERC)

### PR DESCRIPTION
**Pull request by joshuacords**
**Story:** [LPD-28413](https://liferay.atlassian.net/browse/LPD-28413) UserERC
**Subtasks:**
1. [LPD-31076](https://liferay.atlassian.net/browse/LPD-31076)
2. [LPD-32915](https://liferay.atlassian.net/browse/LPD-32915) Test

**Story:** [LPD-28412](https://liferay.atlassian.net/browse/LPD-28412) GroupERC and ScopeGroupERC
**Subtasks:**
1. [LPD-30430](https://liferay.atlassian.net/browse/LPD-30430)
2. [LPD-32914](https://liferay.atlassian.net/browse/LPD-32914) Test

**Author:** @joshuacords

### CODE REVIEW

**Comment by joshuacords** [comment]
ci:test:sf

**Comment by joshuacords** [comment]
ci:test:relevant

**Comment by joshuacords** [comment]
ci:test:search-solr

**Comment by joshuacords** [comment]
ci:test:search-opensearch2

**Comment by joshuacords** [comment]
ci:test:search

File: modules/apps/portal-search/portal-search/src/main/java/com/liferay/portal/search/internal/spi/model/index/contributor/AuditedModelDocumentContributor.java [hunk 1]
```diff
@@ -35,6 +41,8 @@ public void contribute(
 		document.addKeyword(Field.COMPANY_ID, auditedModel.getCompanyId());
 		document.addDate(Field.CREATE_DATE, auditedModel.getCreateDate());
 		document.addDate(Field.MODIFIED_DATE, auditedModel.getModifiedDate());
+		document.addKeyword(
+			"userExternalReferenceCode", _getUserERC(auditedModel));
```
Comment (gustavolimav) [inline modules/apps/portal-search/portal-search/src/main/java/com/liferay/portal/search/internal/spi/model/index/contributor/AuditedModelDocumentContributor.java:45]: I am worried about this database requisition to get external reference codes, it doesn't make sense to hit the database once we have the `auditedModel`. Could we add a `getUserExternalReferenceCode` method to the interface and implement it? would it be to much work?

File: modules/apps/portal-search/portal-search/src/main/java/com/liferay/portal/search/internal/spi/model/index/contributor/GroupedModelDocumentContributor.java [hunk 2]
```diff
@@ -35,15 +40,62 @@ public void contribute(
...
 		GroupedModel groupedModel = (GroupedModel)baseModel;
 
+		long siteGroupId = GroupUtil.getSiteGroupId(
+			groupLocalService, groupedModel.getGroupId());
+
 		document.addKeyword(
-			Field.GROUP_ID,
-			GroupUtil.getSiteGroupId(
-				groupLocalService, groupedModel.getGroupId()));
+			"groupExternalReferenceCode", _getGroupERC(siteGroupId));
```
Comment (gustavolimav) [inline modules/apps/portal-search/portal-search/src/main/java/com/liferay/portal/search/internal/spi/model/index/contributor/GroupedModelDocumentContributor.java:47]: same for `groupedModel` with `groupExternalReferenceCode`

File: modules/apps/portal-search/portal-search/src/main/java/com/liferay/portal/search/internal/spi/model/index/contributor/GroupedModelDocumentContributor.java [hunk 3]
```diff
@@ -35,15 +40,62 @@ public void contribute(
...
+		document.addKeyword(Field.GROUP_ID, siteGroupId);
 
+		document.addKeyword(
+			"scopeGroupExternalReferenceCode", _getScopeGroupERC(groupedModel));
```
Comment (gustavolimav) [inline modules/apps/portal-search/portal-search/src/main/java/com/liferay/portal/search/internal/spi/model/index/contributor/GroupedModelDocumentContributor.java:51]: same for `groupedModel` with `scopeGroupExternalReferenceCode`

File: modules/apps/portal-search/portal-search/src/main/java/com/liferay/portal/search/internal/spi/model/index/contributor/AuditedModelDocumentContributor.java (hunk 1 above)
Comment (gustavolimav) [reply modules/apps/portal-search/portal-search/src/main/java/com/liferay/portal/search/internal/spi/model/index/contributor/AuditedModelDocumentContributor.java:45]: \cc @BryanEngler

**Review by gustavolimav — CHANGES_REQUESTED** [review]
Hey @joshuacords, see my comments.

File: modules/apps/portal-search/portal-search/src/main/java/com/liferay/portal/search/internal/spi/model/index/contributor/AuditedModelDocumentContributor.java (hunk 1 above)
Comment (joshuacords) [reply modules/apps/portal-search/portal-search/src/main/java/com/liferay/portal/search/internal/spi/model/index/contributor/AuditedModelDocumentContributor.java:45]: So adding getUserExternalReferenceCode to the AuditedModel would mean changing 896 classes. Additionally, that alone wouldn't help avoid a database hit as these implementations would still likely be calling the userLocalService (see the [DLFolderModelImpl.getUserUuid](https://github.com/liferay/liferay-portal/blob/4cb0a9753eacfd39efc7f717d22c2bcefb0c7f82/portal-impl/src/com/liferay/portlet/documentlibrary/model/impl/DLFolderModelImpl.java#L656-L665)). To remove the database hit we would need to store the userUuid into each to these models.

Keep in mind that we do cache this information.

**Comment by joshuacords** [comment]
ci:report:55624402

File: modules/apps/portal-search/portal-search/src/main/java/com/liferay/portal/search/internal/spi/model/index/contributor/AuditedModelDocumentContributor.java (hunk 1 above)
Comment (gustavolimav) [reply modules/apps/portal-search/portal-search/src/main/java/com/liferay/portal/search/internal/spi/model/index/contributor/AuditedModelDocumentContributor.java:45]: Thanks for the explanation, I didt  know it was cached. If this is the case, it should be just fine and we wouldn't need to worry about additional database hits. Thx!

**Review by gustavolimav — APPROVED** [review]
LGTM

**Comment by joshuacords** [comment]
ci:forward

**Pull request closed without merging**